### DISCUSSION TITLE
Child or sub logger

### DISCUSSION THREAD

**Discussion by jasonbeach** [gabime/spdlog#2524] [Q&A]
spdlog is pretty fantastic...basically my standard has become every class I write has a spdlog logger in it.  `cout` is basically banned--if it's worth `cout`ing then i's worth sending it to a logger with a console sink. What I'm trying to figure out is the best way to "nest" loggers--i.e. if class A has its logger, class B and C both contain class A as well as have their loggers. I have my own `Logger::get` helper function which queries the global registry for a logger and return it if it's found or creates it if it's not. So each class has:

    LoggerPtr logger_ = Logger::get("Foo");

where the logger's name is set. Behind the scenes when the logger is created, the format is set so the class name (Foo in this case) is printed as part of the log statement. All loggers are going to the same file sink. Unfortunately with this arrangement it's not possible tell which statements came from the class B::A logger vs the class C::A logger. 

I guess instead of setting each logger name statically, I can add a logger "namespace" argument to each class's constructor, but I was really trying to not have any logger arguments in the constructor. I guess there's not really a way to associate the loggers with each without either making them public (they're typically private) or feeding them in through a constructor or some other public function. Just wondering what other people have tried.

**Comment by tt4g** [comment]
How about separating the names of parent and child loggers with a separator (e.g. `/`)?

```cpp
auto parent = spdlog::get("Foo");
std::string logger_name = parent->name() + "/" + child_name;
auto child = parent->clone(logger_name);
```

**Reply by jasonbeach** [reply]
that would probably work...what does the line:
`auto child = parent->close(logger_name);`

do?

**Reply by tt4g** [reply]
Oh... Sorry `close` is typo. `clone` is correct (edited).

**Reply by tt4g** [reply]
`spdlog::logger::clone(std::string)` creates new logger with a new name. It copies the sink and level of the original logger.

https://github.com/gabime/spdlog/blob/6a9d5616712eae9e03570a3efe80dcecba1e3cd0/include/spdlog/logger-inl.h#L157-L163

https://github.com/gabime/spdlog/blob/6a9d5616712eae9e03570a3efe80dcecba1e3cd0/include/spdlog/logger-inl.h#L19-L26
